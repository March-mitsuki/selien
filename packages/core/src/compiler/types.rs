use crate::types::config::Config;

#[derive(Debug)]
pub struct OutputFile {
    pub path: String,
    pub content: String,
}

pub struct Output<'a> {
    pub config: &'a Config,
    pub files: Vec<OutputFile>,
}
