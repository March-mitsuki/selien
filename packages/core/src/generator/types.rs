pub mod ast_type_alias {
    #[derive(Debug, Clone)]
    pub struct StringLiteralNode {
        pub value: String,
    }

    #[derive(Debug, Clone)]
    pub struct NumberLiteralNode {
        pub value: String,
    }

    #[derive(Debug, Clone)]
    pub enum Keywords {
        String,
        Number,
        Boolean,
        Any,
    }

    #[derive(Debug, Clone)]
    pub struct KeywordNode {
        pub value: Keywords,
    }

    #[derive(Debug, Clone)]
    pub struct ArrayNode {
        pub items: Box<Node>,
    }

    #[derive(Debug, Clone)]
    pub struct Property {
        pub identifier: String,
        pub body: Node,
    }

    #[derive(Debug, Clone)]
    pub struct ObjectNode {
        pub values: Vec<Property>,
    }

    #[derive(Debug, Clone)]
    pub struct RefNode {
        pub name: String,
        pub path: String,
    }

    #[derive(Debug, Clone)]
    pub struct DynNode {
        pub name: String,
        pub from: String,
    }

    #[derive(Debug, Clone)]
    pub struct SplitNode {
        pub typescript: Option<Box<Node>>,
        pub go: Option<Box<Node>>,
    }

    #[derive(Debug, Clone)]
    pub enum Node {
        StringLiteral(StringLiteralNode),
        NumberLiteral(NumberLiteralNode),
        Keyword(KeywordNode),
        Array(ArrayNode),
        Object(ObjectNode),
        Ref(RefNode),
        Dyn(DynNode),
        Split(SplitNode),
        Empty,
    }

    impl Node {
        pub fn new() -> Self {
            Self::Empty
        }
    }

    #[derive(Debug)]
    pub struct TypeAliasAst {
        pub identifier: String,
        pub body: Node,
    }
}

pub mod ast_enum {
    #[derive(Debug)]
    pub enum MembersType {
        String,
        Number,
    }

    #[derive(Debug)]
    pub struct Member {
        pub identifier: String,
        pub value: String,
    }

    pub type Members = Vec<Member>;

    #[derive(Debug)]
    pub struct EnumAst {
        pub identifier: String,
        pub r#type: MembersType,
        pub members: Members,
    }
}

#[derive(Debug)]
#[allow(clippy::upper_case_acronyms)]
pub enum AST {
    TypeAlias(ast_type_alias::TypeAliasAst),
    Enum(ast_enum::EnumAst),
}

#[derive(Debug)]
pub struct RefImport {
    pub name: String,
    pub from: String,
}

#[derive(Debug)]
pub struct DynImport {
    pub name: String,
    pub from: String,
}

#[derive(Debug)]
pub enum Import {
    Ref(RefImport),
    Dyn(DynImport),
}

pub type Imports = Vec<Import>;

pub struct TabSize {
    pub go: usize,
    pub typescript: usize,
}

pub const DEFAULT_TABSIZE: TabSize = TabSize {
    go: 4,
    typescript: 2,
};
