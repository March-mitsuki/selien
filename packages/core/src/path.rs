use std::env;
use std::path::{Component, Path, PathBuf};

use log::{error, warn};

/// Process path from input with normalize.
/// If input is relative path, then join it with current working directory.
pub fn process_path<P: AsRef<Path>>(input: P) -> PathBuf {
    let mut path = input.as_ref().to_path_buf();

    if path.is_relative() {
        let cwd = env::current_dir().expect("Can not get current working dir.");
        path = cwd.join(path);
    }

    path = normalize_path(&path);

    path
}

pub fn normalize_path(path: &Path) -> PathBuf {
    let mut components = path.components().peekable();
    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();
        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    for component in components {
        match component {
            Component::Prefix(..) => unreachable!(),
            Component::RootDir => {
                ret.push(component.as_os_str());
            }
            Component::CurDir => {}
            Component::ParentDir => {
                ret.pop();
            }
            Component::Normal(c) => {
                ret.push(c);
            }
        }
    }
    ret
}

/// Convert absolute **selien $ref path** to relative path.
///
/// `e.g.` /rest/user -> rest/user
///
/// This function expected a **unix** path input.
pub fn to_relative(p: &Path) -> PathBuf {
    if !p.starts_with("/") {
        warn!(
            "$ref absolute path should start with '/': {}",
            p.to_str().unwrap()
        );
        return p.to_owned();
    }

    p.strip_prefix("/").unwrap().to_path_buf()
}

/// add dot to relative path.
///
/// `e.g.` home/username/selien -> ./home/username/selien
pub fn add_dot(p: &Path) -> PathBuf {
    // selien $ref abs path should start with '/', even windows
    if p.starts_with("/") {
        error!(
            "$ref absolute path should start with '/': {}",
            p.to_str().unwrap()
        );
        if crate::is_dev() {
            panic!();
        } else {
            std::process::exit(1);
        }
    }

    if std::env::consts::OS == "windows" {
        // use in codegen, so we need to convert windows path to unix path
        PathBuf::from(format!("./{}", p.to_str().unwrap()))
    } else {
        PathBuf::from(".").join(p)
    }
}

pub fn diff_paths<P, B>(path: P, base: B) -> Option<PathBuf>
where
    P: AsRef<Path>,
    B: AsRef<Path>,
{
    let path = path.as_ref();
    let base = base.as_ref();

    if path.is_absolute() != base.is_absolute() {
        if path.is_absolute() {
            Some(PathBuf::from(path))
        } else {
            None
        }
    } else {
        let mut ita = path.components();
        let mut itb = base.components();
        let mut comps: Vec<Component> = vec![];
        loop {
            match (ita.next(), itb.next()) {
                (None, None) => break,
                (Some(a), None) => {
                    comps.push(a);
                    comps.extend(ita.by_ref());
                    break;
                }
                (None, _) => comps.push(Component::ParentDir),
                (Some(a), Some(b)) if comps.is_empty() && a == b => (),
                (Some(a), Some(b)) if b == Component::CurDir => comps.push(a),
                (Some(_), Some(b)) if b == Component::ParentDir => return None,
                (Some(a), Some(_)) => {
                    comps.push(Component::ParentDir);
                    for _ in itb {
                        comps.push(Component::ParentDir);
                    }
                    comps.push(a);
                    comps.extend(ita.by_ref());
                    break;
                }
            }
        }
        Some(comps.iter().map(|c| c.as_os_str()).collect())
    }
}
