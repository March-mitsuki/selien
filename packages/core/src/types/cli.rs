use clap::{arg, command, Args, Parser, Subcommand};

/// Selien a ssot-type-specification compiler.
#[derive(Debug, Parser)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    #[command(subcommand)]
    pub command: Commands,
}

#[derive(Debug, Subcommand)]
pub enum Commands {
    /// Generate code from your spec file.
    /// By default, Selien will generate code for all languages in your config file.
    Gen(GenArgs),
}

#[derive(Debug, Args)]
pub struct GenArgs {
    /// Output language. If not specified, Selien will generate code for all languages in your config file.
    #[arg(short, long)]
    pub output: Option<String>,

    /// Path to your config file. Default is current directory.
    #[arg(short, long, default_value = ".")]
    pub config: String,
}

impl Cli {
    pub fn get_parse() -> Self {
        Self::parse()
    }
}
