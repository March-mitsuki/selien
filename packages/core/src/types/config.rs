use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct Spec {
    pub root: String,
}

#[derive(Debug, Deserialize)]
pub struct OutputGo {
    pub mod_name: String,
    pub root: String,
    pub output: String,
    pub tabsize: usize,
}

#[derive(Debug, Deserialize)]
pub struct OutputTypescript {
    pub output: String,
    pub tabsize: usize,
}

#[derive(Debug, Deserialize)]
pub struct Output {
    pub go: Option<OutputGo>,
    pub typescript: Option<OutputTypescript>,
}

#[derive(Debug, Deserialize)]
pub struct Config {
    pub spec: Spec,
    pub output: Output,
}
