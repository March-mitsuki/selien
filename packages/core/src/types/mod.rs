pub mod cli;
pub mod config;
pub mod lang;
pub mod spec;
