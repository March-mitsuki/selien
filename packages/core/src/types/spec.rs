use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct TypeAliasDef {
    pub identifier: String,
    pub body: serde_yaml::Mapping,
}

#[derive(Debug, Deserialize)]
pub struct EnumDef {
    pub identifier: String,
    pub r#type: String,
    pub members: serde_yaml::Sequence,
}

#[derive(Debug, Deserialize)]
pub enum Def {
    TypeAlias(TypeAliasDef),
    Enum(EnumDef),
}

#[derive(Debug, Deserialize)]
pub struct Spec {
    pub version: String,
    pub path: String,
    pub def: Vec<Def>,
}

impl Spec {
    pub fn new() -> Self {
        Self {
            version: String::new(),
            path: String::new(),
            def: vec![],
        }
    }
}

pub type SpecList = Vec<Spec>;
