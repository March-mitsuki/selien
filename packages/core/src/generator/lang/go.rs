use std::path::PathBuf;

use log::{error, warn};

use super::super::types::{
    ast_enum,
    ast_type_alias::{Node, Property, TypeAliasAst},
    AST,
};
use crate::{
    generator::{
        types::{DynImport, Import, Imports, RefImport},
        utils::capitalize,
    },
    types::lang::SupportedLang,
};

/// Make sure to due with golang imports. (when $ref used)
///
/// golang imports used like this:
/// ```go
/// import (
///    "github.com/yourname/yourpackage"
/// )
///
/// // when use import, must use package name.
/// type YourType struct {
///   YourField yourpackage.YourType
/// }
/// ```
///
/// So we add a **token** when $ref used. Syntax is `[selien-ref]path/to/ref/yourpackage/file.yaml[selien-ref]`
///
/// This token must be replace to actual package name when compile phase.
pub fn generate_go(ast: &AST, imports: &mut Imports, tabsize: usize) -> String {
    match ast {
        AST::Enum(enum_ast) => {
            let mut result = String::from("\n");

            match &enum_ast.r#type {
                ast_enum::MembersType::String => {
                    result += &format!("type {} string\n", capitalize(&enum_ast.identifier));
                }
                ast_enum::MembersType::Number => {
                    result += &format!("type {} int\n", capitalize(&enum_ast.identifier));
                }
            }

            let s = iterate_members(
                &enum_ast.members,
                &enum_ast.r#type,
                &enum_ast.identifier,
                tabsize,
            );
            result += &format!("const (\n{}\n)\n\n", s);

            result
        }
        AST::TypeAlias(type_alias_ast) => {
            let mut result = String::new();
            match &type_alias_ast.body {
                Node::StringLiteral(_) => {
                    let s = format!("type {} string\n", capitalize(&type_alias_ast.identifier));
                    result += &s;
                }
                Node::NumberLiteral(_) => {
                    let s = format!("type {} int\n", capitalize(&type_alias_ast.identifier));
                    result += &s;
                }
                Node::Keyword(node) => {
                    let s = format!(
                        "type {} {}\n",
                        capitalize(&type_alias_ast.identifier),
                        node.value.to_string(SupportedLang::Go)
                    );
                    result += &s;
                }
                Node::Object(node) => {
                    let r: String = node
                        .values
                        .iter()
                        .enumerate()
                        .map(|(idx, v)| {
                            let is_last = idx == node.values.len() - 1;
                            iterate_properties(imports, v, 1, tabsize, is_last)
                        })
                        .collect();
                    result += &format!(
                        "type {} struct {{\n{}\n}}\n",
                        capitalize(&type_alias_ast.identifier),
                        r
                    );
                }
                Node::Array(node) => {
                    let r = iterate_array(imports, &node.items, tabsize);
                    result += &format!("type {} []{}\n", capitalize(&type_alias_ast.identifier), r);
                }
                Node::Ref(node) => {
                    let mut s = format!(
                        "type {} {}\n",
                        capitalize(&type_alias_ast.identifier),
                        capitalize(&node.name)
                    );

                    if !node.path.is_empty() {
                        let ref_token = format!("[selien-ref]{}[selien-ref]", node.path);
                        s = format!(
                            "type {} {}{}\n",
                            capitalize(&type_alias_ast.identifier),
                            ref_token,
                            capitalize(&node.name)
                        );

                        imports.push(Import::Ref(RefImport {
                            name: capitalize(&node.name),
                            from: PathBuf::from(&node.path),
                        }));
                    }

                    result += &s;
                }
                Node::Dyn(node) => {
                    let fp = PathBuf::from(&node.from);
                    let s = format!(
                        "type {} {}.{}\n",
                        capitalize(&type_alias_ast.identifier),
                        fp.file_name().unwrap().to_str().unwrap(),
                        capitalize(&node.name)
                    );
                    imports.push(Import::Dyn(DynImport {
                        name: node.name.clone(),
                        from: node.from.clone(),
                    }));

                    result += &s;
                }
                Node::Union(_) => {
                    let id = capitalize(&type_alias_ast.identifier);
                    warn!(
                        "Union type '{}' using in golang. interface{{}} type will be generated.",
                        &id
                    );
                    let s = format!("type {} interface{{}}\n", id);
                    result += &s;
                }
                Node::Split(split) => {
                    if let Some(node) = &split.go {
                        let s = generate_go(
                            &AST::TypeAlias(TypeAliasAst {
                                identifier: type_alias_ast.identifier.clone(),
                                body: node.as_ref().clone(),
                            }),
                            imports,
                            tabsize,
                        );
                        result += &s;
                    }
                }
                Node::Empty => {
                    error!("Empty node.");
                    if crate::is_dev() {
                        panic!();
                    } else {
                        std::process::exit(1);
                    }
                }
            }
            result
        }
    }
}

fn iterate_properties(
    imports: &mut Imports,
    p: &Property,
    deepth: usize,
    tabsize: usize,
    is_last: bool,
) -> String {
    let indent = " ".repeat(deepth * tabsize);
    let mut result = String::new();
    match &p.body {
        Node::StringLiteral(_) => {
            let mut s = format!(
                "{}{} string `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s;
        }
        Node::NumberLiteral(_) => {
            let mut s = format!(
                "{}{} int `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s;
        }
        Node::Keyword(node) => {
            let mut s = format!(
                "{}{} {} `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                node.value.to_string(SupportedLang::Go),
                &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s;
        }
        Node::Object(node) => {
            let r: String = node
                .values
                .iter()
                .enumerate()
                .map(|(idx, v)| {
                    let is_last = idx == node.values.len() - 1;
                    iterate_properties(imports, v, deepth + 1, tabsize, is_last)
                })
                .collect();
            let mut s = format!(
                "{i}{id} struct {{\n{re}\n{ii}}} `json:\"{j}\"`",
                i = indent,
                id = capitalize(&p.identifier),
                ii = indent,
                re = r,
                j = &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s
        }
        Node::Array(node) => {
            let r = iterate_array(imports, &node.items, tabsize + 1);
            let mut s = format!(
                "{}{} []{} `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                r,
                &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s
        }
        Node::Ref(node) => {
            let mut s = format!(
                "{}{} {} `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                capitalize(&node.name),
                &p.identifier
            );
            if !node.path.is_empty() {
                let ref_token = format!("[selien-ref]{}[selien-ref]", node.path);
                s = format!(
                    "{}{} {}{} `json:\"{}\"`",
                    indent,
                    capitalize(&p.identifier),
                    ref_token,
                    capitalize(&node.name),
                    &p.identifier
                );

                imports.push(Import::Ref(RefImport {
                    name: capitalize(&node.name),
                    from: PathBuf::from(&node.path),
                }));
            }

            if !is_last {
                s += "\n"
            }
            result += &s
        }
        Node::Dyn(node) => {
            let fp = PathBuf::from(&node.from);
            let mut s = format!(
                "{}{} {}.{} `json:\"{}\"`",
                indent,
                capitalize(&p.identifier),
                fp.file_name().unwrap().to_str().unwrap(),
                capitalize(&node.name),
                &p.identifier,
            );
            if !is_last {
                s += "\n"
            }
            imports.push(Import::Dyn(DynImport {
                name: node.name.clone(),
                from: node.from.clone(),
            }));

            result += &s;
        }
        Node::Union(_) => {
            let id = capitalize(&p.identifier);
            warn!(
                "Union type '{}' using in golang object type. interface{{}} type will be generated.",
                &id
            );
            let mut s = format!(
                "{}{} interface{{}} `json:\"{}\"`",
                indent, id, &p.identifier
            );
            if !is_last {
                s += "\n"
            }
            result += &s;
        }
        Node::Split(_) => {
            error!("Split-type can only use on top-level.");
            if crate::is_dev() {
                panic!();
            } else {
                std::process::exit(1);
            }
        }
        Node::Empty => {
            error!("Empty node.");
            if crate::is_dev() {
                panic!();
            } else {
                std::process::exit(1);
            }
        }
    }
    result
}

fn iterate_array(imports: &mut Imports, node: &Node, tabsize: usize) -> String {
    let mut result = String::new();
    match node {
        Node::StringLiteral(_) => result += "string",
        Node::NumberLiteral(_) => result += "int",
        Node::Keyword(node) => result += &node.value.to_string(SupportedLang::Go),
        Node::Object(node) => {
            let r: String = node
                .values
                .iter()
                .enumerate()
                .map(|(idx, v)| {
                    let is_last = idx == node.values.len() - 1;
                    iterate_properties(imports, v, 1, tabsize, is_last)
                })
                .collect();
            result += &format!("struct {{\n{}\n}}", r);
        }
        Node::Array(node) => {
            let r = iterate_array(imports, &node.items, tabsize);
            result += &format!("[]{}", r);
        }
        Node::Ref(node) => {
            let mut s = capitalize(&node.name);
            if !node.path.is_empty() {
                let ref_token = format!("[selien-ref]{}[selien-ref]", node.path);
                s = format!("{}{}", ref_token, capitalize(&node.name));

                imports.push(Import::Ref(RefImport {
                    name: capitalize(&node.name),
                    from: PathBuf::from(&node.path),
                }));
            }

            result += &s
        }
        Node::Dyn(node) => {
            let fp = PathBuf::from(&node.from);
            let s = format!(
                "{}.{}",
                fp.file_name().unwrap().to_str().unwrap(),
                capitalize(&node.name),
            );
            imports.push(Import::Dyn(DynImport {
                name: node.name.clone(),
                from: node.from.clone(),
            }));

            result += &s;
        }
        Node::Union(_) => {
            warn!("Union type using in golang array type. interface{{}} type will be generated.");
            result += "interface{}"
        }
        Node::Split(_) => {
            error!("Split-type can only use on top-level.");
            if crate::is_dev() {
                panic!();
            } else {
                std::process::exit(1);
            }
        }
        Node::Empty => {
            error!("Empty node.");
            if crate::is_dev() {
                panic!();
            } else {
                std::process::exit(1);
            }
        }
    }
    result
}

fn iterate_members(
    members: &ast_enum::Members,
    m_type: &ast_enum::MembersType,
    type_id: &str,
    tabsize: usize,
) -> String {
    let indent = " ".repeat(tabsize);
    let mut result = String::new();

    match m_type {
        ast_enum::MembersType::String => {
            for (idx, m) in members.iter().enumerate() {
                let is_last = idx == members.len() - 1;
                let mut s = format!(
                    "{}{} {} = \"{}\"",
                    indent,
                    capitalize(&m.identifier),
                    capitalize(type_id),
                    m.value
                );
                if !is_last {
                    s += "\n"
                }

                result += &s
            }
        }
        ast_enum::MembersType::Number => {
            for (idx, m) in members.iter().enumerate() {
                let is_last = idx == members.len() - 1;
                let mut s = format!(
                    "{}{} {} = {}",
                    indent,
                    capitalize(&m.identifier),
                    capitalize(type_id),
                    m.value
                );
                if !is_last {
                    s += "\n"
                }

                result += &s
            }
        }
    }

    result
}

// unit test here
#[cfg(test)]
mod test {
    use crate::{
        generator::{lang::for_test, types::DEFAULT_TABSIZE},
        types::lang::SupportedLang,
    };

    #[test]
    fn nested_object() {
        let ast = for_test::nested_object_ast();
        let result = ast.generator(&SupportedLang::Go, &mut vec![], &DEFAULT_TABSIZE);
        insta::assert_yaml_snapshot!(result);
    }

    #[test]
    fn nested_array() {
        let ast = for_test::nested_array_ast();
        let result = ast.generator(&SupportedLang::Go, &mut vec![], &DEFAULT_TABSIZE);
        insta::assert_yaml_snapshot!(result);
    }

    #[test]
    fn object_array() {
        let ast = for_test::object_array_ast();
        let result = ast.generator(&SupportedLang::Go, &mut vec![], &DEFAULT_TABSIZE);
        insta::assert_yaml_snapshot!(result);
    }

    #[test]
    fn string_enum() {
        let ast = for_test::string_enum_ast();
        let result = ast.generator(&SupportedLang::Go, &mut vec![], &DEFAULT_TABSIZE);
        insta::assert_yaml_snapshot!(result);
    }
}
