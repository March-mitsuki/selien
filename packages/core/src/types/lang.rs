use log::error;

use super::config::Config;

#[derive(Debug)]
pub enum SupportedLang {
    TypeScript,
    Go,
}

impl SupportedLang {
    pub fn all() -> Vec<Self> {
        vec![Self::TypeScript, Self::Go]
    }
    pub fn from(s: &str) -> Self {
        let ts_alias = Self::get_alias(&Self::TypeScript);
        let go_alias = Self::get_alias(&Self::Go);
        if ts_alias.contains(&s) {
            return Self::TypeScript;
        }
        if go_alias.contains(&s) {
            return Self::Go;
        }

        error!("Unsupported language or alias: {}.", s);
        if crate::is_dev() {
            panic!();
        } else {
            std::process::exit(1);
        }
    }
    pub fn is_defined_in_config(&self, config: &Config) -> bool {
        match self {
            Self::TypeScript => config.output.typescript.is_some(),
            Self::Go => config.output.go.is_some(),
        }
    }
    pub fn get_alias<'a>(lang: &Self) -> Vec<&'a str> {
        match lang {
            Self::TypeScript => vec!["ts", "typescript"],
            Self::Go => vec!["go", "golang"],
        }
    }
}
