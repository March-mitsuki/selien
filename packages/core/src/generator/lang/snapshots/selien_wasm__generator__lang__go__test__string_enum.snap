---
source: src/generator/lang/go.rs
expression: result
---
"\ntype StringEnum string\nconst (\n    Hello StringEnum = \"hello\"\n    World StringEnum = \"world\"\n)\n\n"

